Use Homebrew curl for audit when specified in the formula

iMichka opened the pull request:
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

MikeMcQuaid commented on Library/Homebrew/dependable.rb:
```diff
@@ -31,6 +31,10 @@ def required?
     !build? && !test? && !optional? && !recommended?
```
```suggestion
    !build? && !test? && !fetch? && !optional? && !recommended?
```

Will also need to fix all the other places that `.test?` is used:

```
m1kebook # rg '\.test\?'                              (master) [~/OSS/Homebrew]
Library/Homebrew/cmd/deps.rb
165:      str = "#{str} [test]" if dep.test?

Library/Homebrew/language/python.rb
139:          next if d.build? || d.test?

Library/Homebrew/os/mac.rb
119:        next false if check_only_runtime_requirements && req.build? && !req.test?

Library/Homebrew/build.rb
55:      if req.prune_from_option?(build) || req.prune_if_build_and_not_dependent?(dependent, formula) || req.test?
66:         (dep.test? && !dep.build?)

Library/Homebrew/dev-cmd/test.rb
67:        next if dependency.test?

Library/Homebrew/dev-cmd/linkage.rb
48:        if args.test?

Library/Homebrew/dependencies_helpers.rb
47:      elsif dep.build? || dep.test?
49:        keep ||= dep.test? && includes.include?("test?") && dependent == root_dependent

Library/Homebrew/formula_installer.rb
539:        keep_build_test ||= req.test? && include_test? && dependent == f
544:           ((req.build? || req.test?) && !keep_build_test) ||
570:      keep_build_test ||= dep.test? && include_test? && @include_test_formulae.include?(dependent.full_name)
574:      if dep.prune_from_option?(build) || ((dep.build? || dep.test?) && !keep_build_test)

Library/Homebrew/test/dependency_expansion_spec.rb
113:      described_class.keep_but_prune_recursive_deps if dep.test?
```

iMichka replied:
It's a little bit "weird" that I have to check all of these (not your fault, the code is as it-is, just some thoughts for the future). I am introducing a new attribute, that's all. I'm fine with the missing `fetch` in the class annotation method, and the missing fetch in the `required?` method. There you are right, they are necessary and I forgot to add them.

But for the others, it feels like a weird design: ideally we should migrate all these boolean logic checks to their respective classes, or refactor code so that a change/check does not feel necessary. Because else it means I can introduce code that will break other code in brew in unexpected ways, and this is bad (if you had not told me, I would have never checked all this code).

I made a few changes by adding fetch when necessary. Can you double check? I am not confortable with the `dependencies_helpers` and `formula_installer` part.

MikeMcQuaid replied:
> It's a little bit "weird" that I have to check all of these (not your fault, the code is as it-is, just some thoughts for the future). I am introducing a new attribute, that's all.

I disagree: you're adding a dependency that you don't actually want to be treated as a dependency in most cases 😁. It makes sense to me that it would need to be special-cased. Also, depending on the case, it may need to behave more like `:build` or `:fetch`.



> Because else it means I can introduce code that will break other code in brew in unexpected ways, and this is bad (if you had not told me, I would have never checked all this code).

I don't think it would break anything, it'll just result in a dependency that's added when it's not necessary.



> I made a few changes by adding fetch when necessary. Can you double check? I am not confortable with the `dependencies_helpers` and `formula_installer` part.

Will take a look!

MikeMcQuaid reviewed (commented):
Nice work so far!

iMichka commented:
Please note that this fixes only the audit, not the fetching.

For example you still get:
```
==> Downloading https://downloads.xiph.org/releases/flac/flac-1.3.3.tar.xz
##O=#  #
curl: (35) error:1400442E:SSL routines:CONNECT_CR_SRVR_HELLO:tlsv1 alert protocol version
Trying a mirror...
==> Downloading https://ftp.osuosl.org/pub/xiph/releases/flac/flac-1.3.3.tar.xz
```

I would propose to fix the fetching in a second pull request (because actually I am struggling to add the brewed curl usage in the curl download strategy, based on the formula's fetch dependencies. I think this can thus be done in a second pull request to not pollute the discussion here.

carlocab commented on Library/Homebrew/utils/curl.rb:
```diff
@@ -75,9 +76,11 @@ def curl_with_workarounds(
         verbose:      verbose,
       }.compact
 
+      curl = use_brewed_curl ? Formula["curl"].opt_bin/"curl" : curl_executable
```
Shouldn't this logic be in the `curl_executable` method?

iMichka replied:
Done

MikeMcQuaid commented on Library/Homebrew/build.rb:
```diff
@@ -52,7 +52,7 @@ def effective_build_options_for(dependent)
     formula.recursive_requirements do |dependent, req|
       build = effective_build_options_for(dependent)
-      if req.prune_from_option?(build) || req.prune_if_build_and_not_dependent?(dependent, formula) || req.test?
+      if req.prune_from_option?(build) || req.prune_if_build_and_not_dependent?(dependent, formula) || req.test? || req.fetch?
```
This maybe needs to be part of `req.prune_if_build_and_not_dependent`

MikeMcQuaid commented on Library/Homebrew/build.rb:
```diff
@@ -63,7 +63,7 @@ def expand_deps
       if dep.prune_from_option?(build) ||
          dep.prune_if_build_and_not_dependent?(dependent, formula) ||
-         (dep.test? && !dep.build?)
+         ((dep.test? || dep.fetch?) && !dep.build?)
```
Makes sense 👍🏻

MikeMcQuaid commented on Library/Homebrew/dev-cmd/test.rb:
```diff
@@ -61,10 +61,11 @@ def test
       end
 
-      # Don't test formulae missing test dependencies
+      # Don't test formulae missing test or fetch dependencies
```
I actually think this can stay as-is, it's fine to test these.

iMichka replied:
Done, I went back to the original version.

MikeMcQuaid commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -571,7 +571,7 @@ def expand_dependencies
                           (formula.head? || !dependent.latest_version_installed?)
 
-      if dep.prune_from_option?(build) || ((dep.build? || dep.test?) && !keep_build_test)
+      if dep.prune_from_option?(build) || ((dep.build? || dep.test? || dep.fetch?) && !keep_build_test)
```
Will need to add something `dep.fetch?` related to `keep_build_test` above

iMichka replied:
Did you mean line 544 in `expand_requirements`? I just added the same logic as for the `expand_dependencies` function.

MikeMcQuaid replied:
I mean something between lines 569-571 in this file. `keep_build_test` is handling whether it's being kept because it's needed for either testing or build. It's not handling the fetch case (or named accordingly) for those lines currently.

MikeMcQuaid commented on Library/Homebrew/os/mac.rb:
```diff
@@ -116,7 +116,7 @@ def sdk_for_formula(f, v = nil, check_only_runtime_requirements: false)
       return Xcode.sdk if f.requirements.any? do |req|
         next false unless req.is_a? XcodeRequirement
-        next false if check_only_runtime_requirements && req.build? && !req.test?
+        next false if check_only_runtime_requirements && req.build? && !req.test? && !req.fetch?
```
Can probably skip this

iMichka replied:
Done, I removed the condition.

iMichka commented:
I'm all done, except for the logic in `expand_deps` and `expand_reqs`. I won't have time until later today to check this, but that part needs more thoughts as the boolean logic in there is quite complex. This might warrant a few tests to be sure we do not break anything.

iMichka commented:
I think I am happy with the changes as-is. I need a new review.

The initial code in `expand_reqs` and `expand_deps` is slightly different, as `expand_deps` has an extra `!dep.build?`. it might be on purpose, or it's missing from `expand_reqs`. I did not investigate that difference further and added the `req/dep.fetch?` condition with the same logic as `req/dep.test?`.

iMichka commented:
If there are no objections, I'll merge this tomorrow.

carlocab reviewed (approved):
Makes sense to me. I wonder if we want something like an `--only-fetch` flag for `brew install`.

nandahkrishna reviewed (approved):
Thanks, Michka!

MikeMcQuaid reviewed (commented):
Sorry for delay here (holiday, moving house), one more comment.

iMichka commented:
- I renamed to dependency type from `:fetch` to `:fetch_and_audit` to make clear that this is used for both. The `fetch` part is not implement in this PR and will be in a second PR. If we want we could also split the dependencies and have a separate `:fetch` and `:audit` type, but I don't think this is needed.

- I introduced an `--include-fetch-and-audit` flag, to be able to write the logic in formula_installer.

- I only implemented the logic for dependencies and ignored the requirements part of the code, as we do not have plans to use fetch/audit requirements for the time being.

iMichka commented:
Right now the `curl` dependency is only installed when `brew install --include-fetch-and-audit <formula>` is used.

We need to change:
- `brew audit` to install the fetch and audit dependency
- `brew fetch` to install the fetch and audit dependency (in a second PR maybe?)
- `brew install` from source to install the fetch and audit dependency

Thoughts? This is getting slightly more complex than expected and I need more help.

nandahkrishna commented:
Thanks for your work here Michka. I'll take a look at some of the remaining tasks this week, it might be better to create separate PRs for them.

> Thoughts? This is getting slightly more complex than expected and I need more help.

One thing to add, if I'm understanding this right, is that we would also need to uninstall Homebrew `curl` (if it's a fetch/audit dependency) after it has been used – once for the fetch and once for the audit, because we're trying to prevent unnecessary linkage with Homebrew `curl`.

Is this thinking correct? If so, I wonder if we should handle that part in test-bot or here. Sorry for the slightly off-topic comment but I thought I'd get it cleared here (or in the issue) before going ahead.

iMichka commented:
> One thing to add, if I'm understanding this right, is that we would also need to uninstall Homebrew curl (if it's a fetch/audit dependency) after it has been used – once for the fetch and once for the audit, because we're trying to prevent unnecessary linkage with Homebrew curl.

Yes, this is what this implementation will be doing: I need to modify both `brew audit` and `brew fetch` to install brew curl, and then uninstall it after it is done (unless it was already installed previously, so we need to keep track of the state of the system). So we will install/uninstall curl twice.

iMichka commented:
Another solution would be to build a portable curl executable, that we could ship to the macOS and Linux CI, and always use that. A little bit like portable ruby.

We did something similar (but not exactly as it is inside a docker image) in the Debian 7 Linux docker image, were we put our own pre-compiled git version in the PATH (see https://github.com/Homebrew/homebrew-linux-dev/blob/master/Dockerfile), because git is too on old on that docker image.

carlocab commented:
This has been suggested previously: https://github.com/Homebrew/install/issues/554#issuecomment-885519531

iMichka commented:
> This has been suggested previously: [Homebrew/install#554 (comment)](https://github.com/Homebrew/install/issues/554#issuecomment-885519531)

Ha yes. @MikeMcQuaid said our version would be hard to keep up to date: I think we would be doing better than the outdated version of Ubuntu 16.04 (and 18.04) soon. We could even auto build a portable curl on each new curl release, using github actions.

carlocab commented on Library/Homebrew/formula_auditor.rb:
```diff
@@ -36,6 +36,7 @@ def initialize(formula, options = {})
       @spdx_license_data = options[:spdx_license_data]
       @spdx_exception_data = options[:spdx_exception_data]
       @tap_audit_exceptions = options[:tap_audit_exceptions]
+      @use_brewed_curl = formula.deps.fetch_and_audit.select { |dep| dep.name == "curl" }.any?
```
```suggestion
      @use_brewed_curl = formula.deps.fetch_and_audit.any? { |dep| dep.name == "curl" }
```
Doesn't this work?

iMichka replied:
Yes it works. Done

iMichka commented:
I implemented the automatic installation of the fetch_and_audit dependents in fetch.rb and audit.rb.

iMichka commented:
The fetch and audit dependencies are now automatically installed when building a formula from source.

iMichka commented:
Everything is now implemented except the last step: the CurlDownLoadStrategy class needs to know about the fetch and audit dependency, so I need to pass the flag to it and implement the logic.

MikeMcQuaid commented on Library/Homebrew/cmd/fetch.rb:
```diff
@@ -86,6 +86,15 @@ def fetch
+        fetch_and_audit_deps.map do |dep|
+          next if fetch_and_audit_deps_installed.include?(dep)
+
+          safe_system ENV["HOMEBREW_BREW_FILE"], "install", dep.name
```
I don't think we should be installing here. I think https://github.com/Homebrew/brew/blob/239bd3cdf79c4e60ddbeb839ddcc53a1837570bc/Library/Homebrew/dependency_collector.rb#L139-L144 

is the right place for this logic as it's what already handles auto-installing these sort of dependencies based on URLs.

Annoyingly, this may also mean that the approach may make more sense for this to be an attribute on the URL e.g. `using: NewCurlDownloadStrategy` or something.

I'm sorry to have not mentioned that earlier, I didn't think about that code until I saw this here.

iMichka replied:
I had a look at this. It is possible to add a `BrewCurlDownLoadStrategy` that uses brew curl. But you need to pass a flag through the whole code down to this place: the entry point being https://github.com/Homebrew/brew/blob/cd900ef71eeecb621faba88aabc641fe93325fe6/Library/Homebrew/resource.rb#L176, i.e. the `url` method from `resource` blocks. This is where it gets tricky, as resources will now have to know if they should used brewed curl or not, based on the formula's `fetch and audit` dependencies. From the first analysis I made this looks like a bigger refactoring or change.

I will make an alternative proposal: the code in `fetch.rb` will install brewed curl based in the `fetch_and_audit` dependency, as already done in this PR. But the final curl call in `curl.rb` can be fixed directly for our issue. I'll push a new version of the PR.

carlocab replied:
Why can't resources that need brewed `curl` also do `using: :brewed_curl` as with any other download strategy?

It might work for the `homepage` too...

MikeMcQuaid replied:
> This is where it gets tricky, as resources will now have to know if they should used brewed curl or not, based on the formula's `fetch and audit` dependencies.

I'm proposing this could eliminate "fetch and audit" dependencies entirely. I'm sorry for the suggested change in approach.

> Why can't resources that need brewed `curl` also do `using: :brewed_curl` as with any other download strategy?

Yeh, this.

> It might work for the `homepage` too...

This would be good, too.

iMichka replied:
I see. I’m working on a solution with :using. Still need to connect some dots but it should work. I’ll update the PR accordingly

MikeMcQuaid commented on Library/Homebrew/cmd/install.rb:
```diff
@@ -204,23 +210,33 @@ def install
+
+      # When building from source, force the installation
+      # of the fetch and audit dependencies.
+      include_fetch_and_audit_formulae = if args.build_from_source_formulae.empty?
```
This also needs to handle building bottles, right?

MikeMcQuaid replied:
🆒 👍🏻

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -182,9 +182,28 @@ def audit
+      fetch_and_audit_deps.map do |dep|
+        next if fetch_and_audit_deps_installed.include?(dep)
+
+        safe_system ENV["HOMEBREW_BREW_FILE"], "install", dep.name
```
Feels fine to install here.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/audit.rb:
```diff
@@ -182,9 +182,28 @@ def audit
+        next if fetch_and_audit_deps_installed.include?(dep)
+
+        safe_system ENV["HOMEBREW_BREW_FILE"],
+                    "uninstall",
```
Don't think this should be uninstalled here. Let `brew test-bot` handle it instead.

iMichka replied:
What happens if you run `brew audit` locally? You might get a different state on you computer than the one left behind by test-bot. Not sure we want to have half of the logic into test-bot, and the other half in brew.

I make sure in the code to check if brewed curl was installed before, and I uninstall it only if it was not the case.

MikeMcQuaid replied:
We don't have anywhere else we're doing "install then uninstall" like this outside of `brew test-bot`. If I'm running `audit` multiple times in a row on relevant it's going to be annoying to have it installed and uninstalled every time.

MikeMcQuaid reviewed (commented):
More comments. Sorry this is taking so long it's just adding a new dependency type like this that's auto installed is inherently going to be hard :( 

I'd be OK given this if we want to hack just the audit and/or fetch case for now until we have a better approach.

MikeMcQuaid commented:
> This has been suggested previously: [Homebrew/install#554 (comment)](https://github.com/Homebrew/install/issues/554#issuecomment-885519531)

I would very strongly like to avoid us maintaining a portable `curl`.

iMichka commented:
I made a new proposal for the `brew fetch`part, please have a look at the change in `curl.rb`. If the result code is 35 and brewed curl is installed, retry the download with brewed curl.

`brew fetch` will still be responsible to install brewed curl, but we use a sort of "indirect" way to hijack the fix into `curl.rb`, without having to rewrite the whole `url`/`resource`/`download strategy` stack.

carlocab commented:
I think we also want to review the `cc` shim and superenv to double check that it's not adding compiler flags or modifying environment variables (e.g. `PATH`) for this new dependency type. I don't think it will, but I think it'd be good to make sure to avoid opportunistic linkage.

Feel free to poke me in the next few days about this if you need me to have a look.

iMichka commented:
Actually I could remove all the `use_brewed_curl` arguments from `formula_auditor.rb` now, as I could use the same logic for audit and fetch: the command will install/uninstall brewed curl, and `curl_with_workaround` will just handle the issue by itself, as implemented now. What do you think?

MikeMcQuaid commented:
> I made a new proposal for the `brew fetch`part, please have a look at the change in `curl.rb`. If the result code is 35 and brewed curl is installed, retry the download with brewed curl.

We had a previous PR for this and: I don't like it, I'm afraid. This is something that should be flagged in individual formulae rather than have magic that reruns `curl` transparently.

iMichka commented:
I re-implemented the change with the `:using` strategy. Not everything is done but you'll see the main idea.

Right now I am struggling to automatically install brewed curl with these lines: https://github.com/Homebrew/brew/pull/11773/files#diff-95bfed356ec6790f13d81d690e41576edf8b9ad6475031f93e74237cbdadb86aR146-R149

To test this, I changed the hello formula:
`url "https://ftp.gnu.org/gnu/hello/hello-2.10.tar.gz", using: :brewed_curl`

And then `brew fetch -s hello` should automatically install brewed curl (make sure you `brew uninstall curl` before).

It does not install curl, which is not what I expected.

iMichka commented:
I'm quite happy with the change now.
The only remaining thing is the automatic installation of brewed curl: I removed the implementation I initially made, and thought https://github.com/Homebrew/brew/pull/11773/files#diff-95bfed356ec6790f13d81d690e41576edf8b9ad6475031f93e74237cbdadb86aR66 would do it. But it does not seem to install curl when doing `brew fetch -s <formula>`. I also tried `brew fetch blastem --head`, and fetching failed because mercurial was not installed: still, `Dependency.new("mercurial", tags)` is added in the download strategy selector, so I am wondering what this does? The install mechanism for these dependencies is maybe broken?

MikeMcQuaid commented:
> I removed the implementation I initially made, and thought https://github.com/Homebrew/brew/pull/11773/files#diff-95bfed356ec6790f13d81d690e41576edf8b9ad6475031f93e74237cbdadb86aR66 would do it.

Does this work for e.g. Subversion etc.? If not, I think it may be fine to leave this without auto-install. If it does, that may be a good place to debug this.

> The install mechanism for these dependencies is maybe broken?

Yes, perhaps. On macOS or Linux?

MikeMcQuaid commented:
@iMichka also: it may be worth comparing `brew install` to `brew fetch` here?

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -559,6 +559,27 @@ def curl(*args, **options)
+      raise CurlDownloadStrategyError, url
+    end
+
+    curl_download resolved_url, to: temporary_path, timeout: timeout, use_brewed_curl: true
```
Is this the only bit that differs from the other `_fetch`? If so, it may be worth having a `_curl_download` or `_curl_download_args` or similar method that can be override to DRY up a bit.

iMichka replied:
Yes it's the only difference. I refactored it as asked;

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -1368,6 +1389,7 @@ def self.detect_from_symbol(symbol)
     when :bzr                    then BazaarDownloadStrategy
     when :svn                    then SubversionDownloadStrategy
     when :curl                   then CurlDownloadStrategy
+    when :brewed_curl            then BrewedCurlDownloadStrategy
```
```suggestion
    when :homebrew_curl            then HomebrewCurlDownloadStrategy
```

Thoughts on this naming, @Homebrew/maintainers? "brewed" is something we only real use in e.g. `HOMEBREW_FORCE_BREWED_CURL`.

carlocab replied:
`:homebrew_curl` is fine too. It's probably a bit clearer for users who are less familiar with Homebrew.

MikeMcQuaid commented on Library/Homebrew/exceptions.rb:
```diff
@@ -602,6 +602,13 @@ def initialize(url)
+# Raised in {BrewedCurlDownloadStrategy#fetch}.
+class BrewedCurlDownloadStrategyError < CurlDownloadStrategyError
+  def initialize(url)
+    super "Brewed curl is not installed for: #{url}"
```
```suggestion
    super "Homebrew-installed `curl` is not installed for: #{url}"
```

MikeMcQuaid commented on Library/Homebrew/utils/curl.rb:
```diff
@@ -14,7 +14,9 @@ module Curl
 
-    def curl_executable
+    def curl_executable(use_brewed_curl: false)
+      return Formula["curl"].opt_bin/"curl" if use_brewed_curl
```
Should this `raise` or similar if this file doesn't exist?

iMichka replied:
I thought about it: I do raise `HomebrewCurlDownloadStrategyError` in resource_auditor and download_strategy. If I want to raise here, I can't use `HomebrewCurlDownloadStrategyError` as it needs to url, and I don't have the url in this method. So I can raise, but the error type will be different, if that is fine for you.

MikeMcQuaid reviewed (commented):
Yeh, way prefer this implementation, great work @iMichka. I'd like to nail down the naming (hopefully within 24h) and then I'm happy with this being merged as-is and we can figure out `brew fetch` issues later.

iMichka commented:
> Does this work for e.g. Subversion etc.? If not, I think it may be fine to leave this without auto-install. If it does, that may be a good place to debug this.

> Yes, perhaps. On macOS or Linux?

On both. We don't have that many svn cases, but you can try `brew install abcl --head`: it will fail with `You must: brew install svn`, though the `svn` dependency has been added. Looks like the return value of these new dependencies is never used.

Not a big deal though right now: my code will fail with `Homebrew-installed `curl` is not installed for <url>` locally and you'll have to install brewed curl manually. And we we will let test-bot install/uninstall brewed curl when necessary (I'll open a second PR for that later).

MikeMcQuaid reviewed (approved):
This looks great! Thanks for sticking with it, nice work @iMichka.

MikeMcQuaid commented:
> On both. We don't have that many svn cases, but you can try `brew install abcl --head`: it will fail with `You must: brew install svn`, though the `svn` dependency has been added. Looks like the return value of these new dependencies is never used.

Interesting. Might be worth an extension (in a new PR) to try and do this automatically. I can take a look at this tomorrow, if requested.

iMichka commented:
Thanks, merging. I will start working on the test-bot change and test this change in homebrew-core.
